Smart grid charging per loadpoint (BC)

naltatis opened the pull request:
fixes #10632
introduces **battery control: grid charging** 🎉

- 🔌 ~~add "cheap/clean grid charging" section in loadpoint modal (if co2 forecast or dyn. prices are configured)~~ (later)
- 🔋 add "cheap/clean grid charging" section in battery settings modal (if battery control is available and see above)
- 🗑️ remove global "smart grid charging" modal (top navi, price/co2 link)

 **Breaking API Change**

- replaced `POST /api/smartcostlimit/:limit` endpoint with `POST /api/loadpoints/:id/smartcostlimit/:limit`. Old Endpoint still exists and writes the limit to all loadpoints.

**Todos**

- [x] per loadpoint & battery UI
- [x] api endpoints
- [x] ~~implement battery forced charging @andig~~ (later) 
- [x] implement per loadpoint enabling @andig 
- [x] publish `smartCostActive` states per lp, ~~battery~~ @andig 
- [x] add e2e tests @naltatis 
- [x] add an `apply to all loadpoints` button to make updating multiple lps easier @naltatis 

**Bonus (maybe separate PRs)**

- add indicator (maybe badge) to loadpoint settings icon to indicate if active slots for current limit exist @naltatis 

Note: Battery charging will be added in a separate PR.

**loadpoint settings / price**

![price per loadpoint](https://github.com/evcc-io/evcc/assets/152287/c5e579ff-78f4-4c16-8fbf-8468e9264441)

**loadpoint settings / co2**

![co2 per loadpoint](https://github.com/evcc-io/evcc/assets/152287/dcb1879c-b75a-4a6b-b62a-b0eb605123b7)

naltatis committed:
Smart grid charging: battery & per loadpoint

naltatis committed:
toml

andig commented:
> POST /api/batterysmartcostlimit/:limit

@naltatis können wir den alten Namen lassen? 

>  publish smartCostActive states per lp, battery @andig

Sonst müssen wir auch hier unterschiedliche Namen nehmen?

andig commented on assets/js/components/BatterySettingsModal.vue:
```diff
@@ -256,22 +257,27 @@
 		prioritySoc: Number,
 		batterySoc: Number,
 		bufferStartSoc: Number,
 		batteryDischargeControl: Boolean,
```
s/batteryDischargeControl/batteryControl/

naltatis replied:
Nope. Dieses Setting sagt weiterhin nur, ob Discharge Lock aktiv ist.

andig commented:
Loadpoint: s/Charging Current/Charger Settings/

naltatis commented:
> > POST /api/batterysmartcostlimit/:limit
> 
> @naltatis können wir den alten Namen lassen?

Auf Site Ebene gibt es ja kein Ladelimit mehr. Würd hier ungern hierarchie einführen.

`batterysmartcostlimit` steuert ja (späterer PR) die Batterie, nicht wie bislang die Ladepunkte. Daher würd ich hier auf jeden Fall nen anderen Endpunkt nehmen. Allein schon um Fehlkonfigurationen (bestehende API Automationen) zu vermeiden.

> > publish smartCostActive states per lp, battery @andig
> 
> Sonst müssen wir auch hier unterschiedliche Namen nehmen?

Wie meinst du das? Aktuell publizieren wir `smartCostActive` an der Seite. Da wir die Limits jetzt ja am LP einstellen müssen wir den Aktiv Zustand auch am jeweiligen LP publizieren. Der Name kann mMn gleich bleiben. Nur der Ort (LP statt site) ändert sich.

andig commented:
> Wie meinst du das?

Konsistenz. Wenn wir an der Site `batterysmartcostlimit` dann sollten wir auch `batterysmartcostactive` publishen.

naltatis commented:
> > Wie meinst du das?
> 
> Konsistenz. Wenn wir an der Site `batterysmartcostlimit` dann sollten wir auch `batterysmartcostactive` publishen.

Achso. Du meintest das für Battery. Ja auf jeden Fall.

naltatis committed:
revert battery smart cost (for now), fix tests and lint

naltatis committed:
generate

naltatis commented:
@andig ich hab hier jetzt mal aufgeräumt und alles batteriespezifische wie besprochen vorerst reverted.

naltatis committed:
Merge branch 'master' into feature/costlimit_per_lp_and_battery

naltatis committed:
chore: upgrade npm dependencies (#12207)

StefanSchoof committed:
chore: disable blank issues and add discussions link (#12245)

andig committed:
chore: cleanup Menneckes naming (#12202)

andig committed:
Sunspec: Fix regression in model selection (#12274)

andig committed:
Add Tesla proxy client (#12174)

naltatis committed:
Custom select: add button semantics (#12279)

naltatis committed:
Visualization: hide values if not enough space (#12280)

naltatis committed:
Config UI: move generic meters to top (#12281)

naltatis committed:
Plan UI: debounce preview (#12284)

andig committed:
chore: better linter output

andig committed:
chore: use go 1.22 (#11007)

andig committed:
Tronity: improve status detection (#12296)

berndkrannich committed:
Fronius Solar API: add descriptions (#12310)

berndkrannich committed:
chore: don't set mask here as it is the default (#12313)

Signed-off-by: berndkrannich <github@krannich.net>

GrimmiMeloni committed:
Easee: always adjust DCC after resuming charge (#12258)

andig committed:
Tesla: unify implementations (BC) (#12277)

berndkrannich committed:
Fronius Solar API: explain limitations of active battery control (#12322)

Signed-off-by: berndkrannich <github@krannich.net>

andig committed:
Tesla: fix build

andig committed:
Powerwall: revert unification with tesla identity

andig committed:
Tesla: fix vehicle asleep detection

creuter27 commented:
Hallo @naltatis ,
Sehe ich das richtig dass dieses Feature derzeit nur im branch costlimit_per_lp implementiert ist und ich meine Installation auf diesen branch switchen müsste um das nutzen zu können?
Wird das demnächst in den master branch übernommen werden?

Noch eine Frage: Wenn ich den Hausakku genauso steuern möchte wie die Vehicle-Akkus (also einen bestimmten SOC zu einem Termin vorgeben und evvcc die optimale Zeit ermitteln lassen) dann ist die einzige Möglichkeit über die Custom-Konfiguration einen "Dummy-Loadpoint" und ein "Dummy-Vehicle" zu erstellen richtig? Oder habe ich da etwas übersehen?

cschlipf committed:
Sofarsolar-g3: remove delay (#12341)

andig committed:
chore: fix logical order

andig committed:
chore: add test

naltatis committed:
Battery settings: improve buffer start

naltatis committed:
initial priority soc fix

andig committed:
Tesla: fix detecting vehicle asleep (#12369)

naltatis committed:
Publish as mDNS service

naltatis committed:
clean

andig committed:
Tesla: always expose the controller interface (#12391)

andig committed:
Add support for Amber Electric (AU) (#12381)

premultiply committed:
Bender: add Spelsberg and Cubos (#12405)

andig committed:
Modbus: fix int to float conversion for int setter

weblate committed:
Translations update from Hosted Weblate (#12299)

* Translated using Weblate (Dutch)

Currently translated at 100.0% (358 of 358 strings)

Co-authored-by: hyperbart <registeradressbart@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/nl/
Translation: evcc/evcc

* Translated using Weblate (Czech)

Currently translated at 36.8% (132 of 358 strings)

Co-authored-by: czkanec <facebook@janecka.info>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/cs/
Translation: evcc/evcc

* Translated using Weblate (Romanian)

Currently translated at 74.3% (266 of 358 strings)

Co-authored-by: infinus <azskylineGTR@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/ro/
Translation: evcc/evcc

* Tesla: fix vehicle asleep detection

* fix toml

---------

Co-authored-by: hyperbart <registeradressbart@gmail.com>
Co-authored-by: czkanec <facebook@janecka.info>
Co-authored-by: infinus <azskylineGTR@gmail.com>
Co-authored-by: andig <cpuidle@gmx.de>
Co-authored-by: premultiply <4681172+premultiply@users.noreply.github.com>

andig committed:
SolarEdge: re-add battery control (#11904)

nicok1 committed:
Offer rollback option for interactive upgrading  (#11771)

chrissooo committed:
Retry vehicle wake-up if necessary (#11707)

andig committed:
Add logic

andig committed:
wip

andig committed:
Implement limit

andig committed:
Add mqtt api

andig committed:
cli: allow setting vehicle current (if supported) (#12429)

andig committed:
chore: simplify how mqtt setters are defined (#12428)

andig committed:
chore: upgrade libraries

naltatis committed:
Merge branch 'master' into feature/costlimit_per_lp_and_battery

naltatis committed:
merge conflicts

naltatis committed:
generate

andig commented on core/site.go:
```diff
@@ -789,6 +781,7 @@ func (site *Site) update(lp Updater) {
 		flexiblePower = site.prioritizer.GetChargePowerFlexibility(lp)
 	}
 
+	// TODO: implement active check for loadpoints and battery
```
@naltatis was meinst Du hiermit?

naltatis replied:
Publishing von `smartCostActive`. Hast hast du am LP ja schon gemacht. Battery ist ausgeklammert. Daher würd ich sagen: ✅

naltatis committed:
add e2e tests

naltatis committed:
i18n labels

naltatis committed:
clean

andig committed:
Align and simplify handler definition

andig committed:
Set smart cost limit globally

naltatis committed:
remove deprecated tests

naltatis committed:
apply to all

naltatis commented:
Added "Apply everywhere?" Button and save feedback.

![Bildschirmfoto 2024-02-24 um 17 39 46](https://github.com/evcc-io/evcc/assets/152287/b8801841-1825-461e-b51c-470160d0220e)

![Bildschirmfoto 2024-02-24 um 17 39 54](https://github.com/evcc-io/evcc/assets/152287/80888547-e8e0-4273-a224-a34bb435d8a8)

![Bildschirmfoto 2024-02-24 um 17 40 00](https://github.com/evcc-io/evcc/assets/152287/06a6ba37-104b-4abc-a2d5-d9c8feca0e69)

naltatis commented:
@andig wenn du keine Anmerkungen mehr hast kanns rein.

andig commented:
Flupp

VolkerK62 commented:
@naltatis   "Apply everywhere" ist im Dunkel Modus nicht sichtbar. Wenn man es weiß, kann man es anklicken ;)
![image](https://github.com/evcc-io/evcc/assets/99199865/d0aa3c76-606b-47c0-8526-29dd28f2cbc3)
